### DISCUSSION TITLE
Could not embed because more than one relationship was found

### DISCUSSION THREAD

**Discussion by Electric-Dragon** [supabase/supabase#6274] [Questions] [answered]
I am trying to query a foreign table, but I get the following error:
`Could not embed because more than one relationship was found for 'adaptation_notifications' and 'series'`

There are the following tables: 

---

`adaptation_notifications`

<img width="1123" alt="an" src="https://user-images.githubusercontent.com/45729391/161414158-356b31ef-e0b3-4ea6-80a5-98e658a1048b.png">

The series column references the `id` column on the `series` table

---

This is an `embedded view` that references the `id` column on the `series` table as well

`series_popularity`

<img width="357" alt="sp" src="https://user-images.githubusercontent.com/45729391/161414222-6c1e9b09-1145-4811-bf84-6c4b331320c9.png">

---

This is the query:

```
      const {data: notifications, error: error_} = await supabase
        .from('adaptation_notifications')
        .select('id,from(id,username),to(id,username),series(id,title,novel),status,when')
        .or(`from.eq.${user.id},to.eq.${user.id}`)
        .order('when', { ascending: false })
```

Edit: I want to access the data from the `series` table and not the `view`

**Comment by steve-chavez** [comment, marked as answer]
Could you paste the full error? The `details` and `hint` fields should reveal more.

**Reply by Electric-Dragon** [reply]
That's the only error I get... where do I get the hint and details from?

**Reply by steve-chavez** [reply]
Hm, are you using the supabase-js client?

**Reply by Electric-Dragon** [reply]
Yes

**Reply by steve-chavez** [reply]
If you do a `console.log(error)` it should show the full error(like shown here https://github.com/supabase/supabase/discussions/2579)

**Reply by Electric-Dragon** [reply]
<img width="1440" alt="Bildschirmfoto 2022-04-03 um 4 05 52 PM" src="https://user-images.githubusercontent.com/45729391/161423506-35d54877-60ad-469d-a864-66326c717364.png">
<img width="1440" alt="Bildschirmfoto 2022-04-03 um 4 05 23 PM" src="https://user-images.githubusercontent.com/45729391/161423508-4bf5e0cc-a72f-464c-abd7-199ad2e564c6.png">

**Reply by Electric-Dragon** [reply]
I tried it now with `.select('id,from(id,username),to(id,username),series!adaptation_notifications_series_fkey(id,title,novel),status,when')` but it still doesn't work

**Reply by Electric-Dragon** [reply]
@steve-chavez I now tried it with `.select('id,from(id,username),to(id,username),series!adaptation_notifications_series_fkey[series][id,title,novel],status,when')`

The error goes away, but the `series property` only returns the `id` and not `title` or `novel`

**Reply by steve-chavez** [reply]
@Electric-Dragon According to the error there is another table `series` that has a relationship with `adaptation_notifications`, can you share the definition of `series`?

**Reply by Electric-Dragon** [reply]
@steve-chavez 

This is the table definition

```
    id uuid NOT NULL DEFAULT uuid_generate_v4(),
    title character varying(50) COLLATE pg_catalog."default" NOT NULL,
    summary character varying(500) COLLATE pg_catalog."default" NOT NULL,
    chapcount integer NOT NULL DEFAULT 0,
    cover text COLLATE pg_catalog."default",
    createdat timestamp without time zone NOT NULL DEFAULT CURRENT_TIMESTAMP,
    updatedat timestamp without time zone NOT NULL DEFAULT CURRENT_TIMESTAMP,
    status character varying(1) COLLATE pg_catalog."default" NOT NULL DEFAULT 'd'::character varying,
    adaptation uuid,
    creator uuid NOT NULL,
    novel boolean NOT NULL,
    mature boolean NOT NULL,
    publicchapcount integer NOT NULL DEFAULT 0,
    genre1 character varying COLLATE pg_catalog."default" NOT NULL,
    genre2 character varying COLLATE pg_catalog."default" NOT NULL,
    comment_count integer NOT NULL DEFAULT 0,
    fts tsvector GENERATED ALWAYS AS (to_tsvector('english'::regconfig, (((summary)::text || ' '::text) || (title)::text))) STORED,
    total_read_count integer NOT NULL DEFAULT 0,
    CONSTRAINT series_pkey PRIMARY KEY (id),
    CONSTRAINT series_adaptation_fkey FOREIGN KEY (adaptation)
        REFERENCES public.series (id) MATCH SIMPLE
        ON UPDATE NO ACTION
        ON DELETE NO ACTION,
    CONSTRAINT series_creator_fkey FOREIGN KEY (creator)
        REFERENCES public.public_profile (id) MATCH SIMPLE
        ON UPDATE NO ACTION
        ON DELETE NO ACTION
```

**Reply by steve-chavez** [reply]
Can you try it like:

```js
      const {data: notifications, error: error_} = await supabase
        .from('adaptation_notifications')
        .select('id,from(id,username),to(id,username),series!series(id,title,novel),status,when')
        .or(`from.eq.${user.id},to.eq.${user.id}`)
        .order('when', { ascending: false })
```

Basically using `series!series(*)`.

**Reply by Electric-Dragon** [reply]
Nope, it still gives me the same error: `Could not embed because more than one relationship was found for 'adaptation_notifications' and 'series'`

Hint: `Try changing 'series' to one of the following: 'series!adaptation_notifications_series_fkey', 'series_popularity!adaptation_notifications_series_fkey'. Find the desired relationship in the 'details' key.`

Details: 
```
[{relationship: "adaptation_notifications_series_fkey[series][id]", embedding: "adaptation_notifications with series", cardinality: "many-to-one"},
 {relationship: "adaptation_notifications_series_fkey[series][id]", embedding: "adaptation_notifications with series_popularity", cardinality: "many-to-one"}]
```

**Reply by steve-chavez** [reply]
@Electric-Dragon Just to confirm, can you embed `series_popularity` like the below?

```js
      const {data: notifications, error: error_} = await supabase
        .from('adaptation_notifications')
        .select('id,from(id,username),to(id,username),series_popularity(*),status,when')
        .or(`from.eq.${user.id},to.eq.${user.id}`)
        .order('when', { ascending: false })
```

**Reply by steve-chavez** [reply]
Also, please share the `adaptation_notifications` table definition.

**Reply by Electric-Dragon** [reply]
@steve-chavez Yes `series_popularity` is embedded successfully

<img width="1440" alt="Bildschirmfoto 2022-04-04 um 11 51 22 PM" src="https://user-images.githubusercontent.com/45729391/161607290-abd9255a-effa-4dc2-8f86-6cd58575bbdb.png">


Here is the `series_popularity` definition. It is a `view`

```
CREATE OR REPLACE VIEW public.series_popularity
 AS
 SELECT series.id,
    0.8 * get_series_follows(series.id)::numeric + 0.2 * series.total_read_count::numeric AS popularity_score
   FROM series;

```

**Reply by steve-chavez** [reply]
Ok, cool. Please share the `adaptation_notifications` table definition, I think its foreign keys names are what's needed to resolve this.

**Reply by steve-chavez** [reply]
So far, the problem seems to be you have a `series` column and a `series` table. If the column were to be renamed to something like `target_series` your original example would work.

**Reply by Electric-Dragon** [reply]
@steve-chavez Here is `adaptation_notifications`

```
CREATE TABLE IF NOT EXISTS public.adaptation_notifications
(
    id bigint NOT NULL GENERATED BY DEFAULT AS IDENTITY ( INCREMENT 1 START 1 MINVALUE 1 MAXVALUE 9223372036854775807 CACHE 1 ),
    "from" uuid NOT NULL,
    "to" uuid NOT NULL,
    series uuid,
    status text COLLATE pg_catalog."default" NOT NULL,
    "when" timestamp without time zone NOT NULL DEFAULT now(),
    CONSTRAINT adaptation_notifications_pkey PRIMARY KEY (id),
    CONSTRAINT adaptation_notifications_from_fkey FOREIGN KEY ("from")
        REFERENCES public.public_profile (id) MATCH SIMPLE
        ON UPDATE NO ACTION
        ON DELETE NO ACTION,
    CONSTRAINT adaptation_notifications_series_fkey FOREIGN KEY (series)
        REFERENCES public.series (id) MATCH SIMPLE
        ON UPDATE NO ACTION
        ON DELETE NO ACTION,
    CONSTRAINT adaptation_notifications_to_fkey FOREIGN KEY ("to")
        REFERENCES public.public_profile (id) MATCH SIMPLE
        ON UPDATE NO ACTION
        ON DELETE NO ACTION
)
```

Please let me know if renaming the column is the only thing I can do.

**Reply by steve-chavez** [reply]
@Electric-Dragon Try this one:

```js
      const {data: notifications, error: error_} = await supabase
        .from('adaptation_notifications')
        .select('id,from(id,username),to(id,username),series!adaptation_notifications_series_fkey(*),status,when')
        .or(`from.eq.${user.id},to.eq.${user.id}`)
        .order('when', { ascending: false })
```

Basically using `series!adaptation_notifications_series_fkey(*)` now.

**Reply by Electric-Dragon** [reply]
Nope, it still gives me the same error: `Could not embed because more than one relationship was found for 'adaptation_notifications' and 'series'`

Hint: `Try changing 'series' to one of the following: 'series!adaptation_notifications_series_fkey', 'series_popularity!adaptation_notifications_series_fkey'. Find the desired relationship in the 'details' key.`

Details:
```
{relationship: "adaptation_notifications_series_fkey[series][id]", embedding: "adaptation_notifications with series", cardinality: "many-to-one"},
 {relationship: "adaptation_notifications_series_fkey[series][id]", embedding: "adaptation_notifications with series_popularity", cardinality: "many-to-one"}] 
```

The query works when I use this, but only the `series id` is returned
```
const {data: notifications, error: error_} = await supabase
        .from('adaptation_notifications')
        .select('id,from(id,username),to(id,username),series!adaptation_notifications_series_fkey[series](id,title,novel),status,when')
        .or(`from.eq.${user.id},to.eq.${user.id}`)
        .order('when', { ascending: false })
```

**Reply by steve-chavez** [reply]
> series!adaptation_notifications_series_fkey[series]

Hm, the brackets `[]` are not valid syntax actually, seems that hint might be getting ignored.

@Electric-Dragon Can you share me your project ref so I can debug this further? If you don't want to share it publicly you can send me the ref at steve@supabase.io.

**Reply by Electric-Dragon** [reply]
@steve-chavez I have done the needful

**Reply by steve-chavez** [reply]
Thanks a lot @Electric-Dragon. Sorry, it seems you have hit a limitation in how PostgREST does disambiguation when embedding.

Will report this as a bug in postgrest issue tracker.

Not ideal, but as a workaround, you can rename the `series` column as I suggested above in https://github.com/supabase/supabase/discussions/6274#discussioncomment-2503590

**Reply by Electric-Dragon** [reply]
Okay...so I rename the `series` column in the `adaptation_notifications` and then I run it with `target_series!series(*)`?

**Reply by steve-chavez** [reply]
Yes, that should do it. Let me know how it goes.

**Reply by Electric-Dragon** [reply]
@steve-chavez It now gives me an error saying `Could not find a relationship between 'adaptation_notifications' and 'target_series' in the schema cache`

**Reply by steve-chavez** [reply]
@Electric-Dragon Actually you can now use plain `series` to do the embedding, like in your original snippet:

```js
      const {data: notifications, error: error_} = await supabase
        .from('adaptation_notifications')
        .select('id,from(id,username),to(id,username),series(id,title,novel),status,when')
        .or(`from.eq.${user.id},to.eq.${user.id}`)
        .order('when', { ascending: false })
```

**Reply by Electric-Dragon** [reply]
Yes this worked...but can you tell me why?

**Reply by steve-chavez** [reply]
Sure. Now the `series(*)` unambiguously referred to the `series` table, unlike before when there was a `series` column in the `adaptation_notifications` table which guided the embedding to be done over this column - and this column referred(through foreign key constraints) to the `series` table or the `series_popularity` view([PostgREST infers FKs on views based on its source tables](https://postgrest.org/en/latest/api.html#embedding-views)).

**Reply by Electric-Dragon** [reply]
Thanks a lot!

**Reply by steve-chavez** [reply]
Thanks a lot for your patience here @Electric-Dragon. We'll definitely revise this - I think just automatically trying to embed on views based on a column is not a good default.

**Reply by Electric-Dragon** [reply]
Yes...especially for beginners in Postgres😅

**Comment by Electric-Dragon** [comment]
@GaryAustin1 could you help me out?

**Reply by GaryAustin1** [reply]
Steve is the main PostgREST dev.  I might be able to dig into it later but not something I can see without doing some research.

**Reply by Electric-Dragon** [reply]
Oh okay...I have received help from you before, so I thought about mentioning u as well. Thanks for the info tho!
